### PR TITLE
[Test] add instructions to the contract methods

### PR DESCRIPTION
**Pull request by Jim8y**
This pr disassemble the script and add corresponding instructions to the artifacts. 

This can be very useful to add new SystemCalls by implementing them in contract and compile to get instructions directly.

Its getting harder and harder to implement system methods, now with this pr, it can be done eaisly and directly in C#.

### CODE REVIEW

**Comment by Jim8y** [comment]
Core part of the disassember is copied from Optimizer implemented by @Hecate2 but i have to move and adapt it into a seperate project due to cycle dependency issue.

File: src/Neo.Disassembler.CSharp/Disassembler.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    private static readonly Regex RangeRegex = new(@"(\d+)\-(\d+)", RegexOptions.Compiled);
+
+    public static List<Instruction> ConvertScriptToInstructions(byte[] script)
+    {
+        var res = EnumerateInstructions(script);
+
+        return res.Select(x => x.instruction).ToList();
+    }
+
+    public static List<(int address, Instruction instruction)> ConvertMethodToInstructions(NefFile nef, JToken DebugInfo, string method)
```
Comment (Jim8y) [inline src/Neo.Disassembler.CSharp/Disassembler.cs:22]: this is the core part of this pr

**Review by shargon — COMMENTED** [review]
Is not better to add the comment above the method, inside the summary or in a different  tag?

File: src/Neo.Disassembler.CSharp/Disassembler.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        var address = 0;
+        var opcode = OpCode.PUSH0;
+        Instruction instruction;
+        for (; address < script.Length; address += instruction.Size)
+        {
+            instruction = script.GetInstruction(address);
+            opcode = instruction.OpCode;
+            yield return (address, instruction);
+        }
+        if (opcode != OpCode.RET)
```
Comment (shargon) [inline src/Neo.Disassembler.CSharp/Disassembler.cs:57]: Technically this is out of the script

**Comment by Jim8y** [comment]
> Is not better to add the comment above the method, inside the summary or in a different tag?

I thought about that as well, but you have comments and artibute above method.

**Comment by shargon** [comment]
```
/// <summary>
/// This summary.
/// </summary>
/// <remarks>
/// <---- HERE?
/// </remarks>
public void MyMethod()
{
    // Method implementation
}
```

**Comment by Jim8y** [comment]
> ```
> /// <summary>
> /// This summary.
> /// </summary>
> /// <remarks>
> /// <---- HERE?
> /// </remarks>
> public void MyMethod()
> {
>     // Method implementation
> }
> ```
> 
> @Jim8y what do you think?

as long as it works, accept your suggestion as usual.

File: src/Neo.Disassembler.CSharp/Disassembler.cs (hunk 2 above)
Comment (shargon) [reply src/Neo.Disassembler.CSharp/Disassembler.cs:57]: Some of them end's with ret, and other don't, I think this line should be removed

Comment (Jim8y) [reply src/Neo.Disassembler.CSharp/Disassembler.cs:57]: > Some of them end's with ret, and other don't, I think this line should be removed

I will check it, this code was adapted from the optimizer

Comment (Jim8y) [reply src/Neo.Disassembler.CSharp/Disassembler.cs:59]:
```C#
        for (; address < script.Length; address += instruction.Size)
        {
            instruction = script.GetInstruction(address);
            opcode = instruction.OpCode;
            yield return (address, instruction);
        }
```

will process instructions already, without caring about Ret. 

```C#
        if (opcode != OpCode.RET)
            yield return (address, Instruction.RET);
```
might be for ensuring the complete parsing.

**Review by shargon — APPROVED** [review]

**Review by shargon — DISMISSED** [review]

File: src/Neo.SmartContract.Testing/Extensions/ArtifactExtensions.cs [hunk 3]
```diff
@@ -36,10 +102,9 @@ public static class ArtifactExtensions
         /// <param name="nef">Nef file</param>
         /// <param name="generateProperties">Generate properties</param>
         /// <returns>Source</returns>
-        public static string GetArtifactsSource(this ContractManifest manifest, string? name = null, NefFile? nef = null, bool generateProperties = true)
+        public static string GetArtifactsSource(this ContractManifest manifest, string? name, NefFile? nef = null, bool generateProperties = true, JToken? debugInfo = null)
```
Comment (shargon) [inline src/Neo.SmartContract.Testing/Extensions/ArtifactExtensions.cs:105]:
```suggestion
        public static string GetArtifactsSource(this ContractManifest manifest, string? name = null, NefFile? nef = null, bool generateProperties = true, JToken? debugInfo = null)
```

**Pull request merged**